spec/provider/bintray_spec.rb

```diff
@@ -0,0 +1,261 @@
+
+      init_provider(provider)
+      request_details = provider.publish_version
+      expect(request_details.get_path).to eq("/content/#{subject}/#{repo}/#{package_name}/#{version_name}/publish")
+    end
+  end
+
+  describe "gpg_sign_version_without_passphrase" do
+    example do
+      descriptor = JSON.parse(get_descriptor_content)
+      package = descriptor["package"]
+      package_name = package["name"]
+      subject = package["subject"]
+      repo = package["repo"]
+      version_name = descriptor["version"]["name"]
+
+      init_provider(provider)
+      request_details = provider.gpg_sign_version
+      expect(request_details.get_path).to eq("/gpg/#{subject}/#{repo}/#{package_name}/versions/#{version_name}")
+      expect(request_details.get_body).to eq(nil)
+    end
+  end
+
+  describe "gpg_sign_version_with_passphrase" do
+    example do
+      descriptor = JSON.parse(get_descriptor_content)
+      package = descriptor["package"]
+      package_name = package["name"]
+      subject = package["subject"]
+      repo = package["repo"]
+      version_name = descriptor["version"]["name"]
+
+      init_provider(provider_with_passphrase)
+      request_details = provider_with_passphrase.gpg_sign_version
+      expect(request_details.get_path).to eq("/gpg/#{subject}/#{repo}/#{package_name}/versions/#{version_name}")
+
+      body = {
+          'passphrase' => 'passphrase'
+      }
+      expect(request_details.get_body).to eq(body)
+    end
+  end
+
+  describe "upload_files" do
+    example do
+      init_provider(provider)
+      files_to_upload = Hash.new()
+      matrix_params = {
+          'p1' => 'a',
+          'p2' => 'b'
+      }
+
+      provider.add_if_matches(files_to_upload, 'build/files/a.gem', 'build/files/(.*.gem)', 'exclude', 'a/b/$1', nil)
+      provider.add_if_matches(files_to_upload, 'build/files/b.gem', 'build/files/(.*.gem)', nil, 'a/b/$1', matrix_params)
+      provider.add_if_matches(files_to_upload, 'build/files/c.gem', 'build/files/(.*.gem)', '.*files.*', 'a/b/$1', nil)
+      provider.add_if_matches(files_to_upload, 'build/files/c.txt', 'build/files/(.*.gem)', 'exclude', 'a/b/$1', nil)
+
+      expect(files_to_upload["build/files/a.gem"]).not_to eq(nil)
+      expect(files_to_upload["build/files/a.gem"].get_upload_path).to eq('a/b/a.gem')
+      expect(files_to_upload["build/files/a.gem"].get_matrix_params).to eq(nil)
+
+      expect(files_to_upload["build/files/b.gem"]).not_to eq(nil)
+      expect(files_to_upload["build/files/b.gem"].get_upload_path).to eq('a/b/b.gem')
+      expect(files_to_upload["build/files/b.gem"].get_matrix_params).to eq(matrix_params)
+
+      expect(files_to_upload["build/files/c.gem"]).to eq(nil)
+      expect(files_to_upload["build/files/c.txt"]).to eq(nil)
+    end
+  end
+
+  def init_provider(bintray)
+    bintray.init_from_args
+    bintray.set_descriptor(get_descriptor_content)
+    bintray.set_test_mode
+  end
+
+  def get_descriptor_content
+    return '{'\
```

Comment:
For future reference, Ruby supports multiline strings with the `"""` denotation, e.g.:

```
2.1.5 :001 > foo = """
2.1.5 :002"> this
2.1.5 :003"> is  
2.1.5 :004"> sparta!
2.1.5 :005"> """
 => "\nthis\nis\nsparta!\n" 
2.1.5 :006 > puts foo

this
is
sparta!
 => nil 
2.1.5 :007 > 
``